x/programs/cmd/simulator/src/lib.rs

```diff
@@ -227,105 +239,70 @@ impl Client {
-where
-    P: AsRef<OsStr>,
-{
-    let mut child = Command::new(path)
-        .arg("run")
-        .arg("--cleanup")
-        .arg("-")
-        .stdin(Stdio::piped())
-        .stdout(Stdio::piped())
-        .spawn()?;
-
-    // write json to stdin
-    let input =
-        serde_json::to_string(plan).map_err(|e| format!("failed to serialize json: {e}"))?;
-    if let Some(ref mut stdin) = child.stdin {
-        stdin
-            .write_all(input.as_bytes())
-            .map_err(|e| format!("failed to write to stdin: {e}"))?;
+impl Default for Client {
+    fn default() -> Self {
+        Self::new()
     }
-
-    child
-        .wait_with_output()
-        .map_err(|e| format!("failed to wait for child: {e}").into())
 }
 
-fn run_steps<P, T>(path: P, plan: &Plan) -> Result<Vec<T>, Box<dyn Error>>
-where
-    T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
-{
-    let output = cmd_output(path, plan)?;
-    let mut items: Vec<T> = Vec::new();
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
-    }
-
-    for line in String::from_utf8(output.stdout)?
-        .lines()
-        .filter(|line| !line.trim().is_empty())
-    {
-        let item = serde_json::from_str(line)
-            .map_err(|e| format!("failed to parse output to json: {e}"))?;
-        items.push(item);
-    }
-
-    Ok(items)
+struct SimulatorChild<W, R> {
+    writer: W,
+    responses: R,
 }
 
-fn run_step<P, T>(path: P, plan: &Plan) -> Result<T, Box<dyn Error>>
+impl<W, R> SimulatorChild<W, R>
 where
-    T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
+    W: Write,
+    R: Iterator<Item = Result<PlanResponse, ClientError>>,
 {
-    let output = cmd_output(path, plan)?;
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
+    fn run_step(&mut self, caller_key: &String, step: &Step) -> Result<PlanResponse, ClientError> {
+        let run_command = b"run --step '";
+        self.writer.write_all(run_command)?;
+
+        let step = SimulatorStep { caller_key, step };
+        let input = serde_json::to_vec(&step)?;
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        let resp = self
+            .responses
+            .next()
+            .ok_or(ClientError::Custom("EOF".into()))??;
```

Comment:
You should make `EOF` a variant of `ClientError` instead of unnecessarily allocating a `String` here.

Reply:
Done, thanks for the below further explanation !